### PR TITLE
[Streaming]Streaming support metric reporter

### PR DESCRIPTION
**Pull request by ashione**
Streaming supports stats module for perf profiling.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: streaming/src/runtime_context.h [hunk 1]
```diff
@@ -30,9 +33,36 @@ class RuntimeContext {
...
 
+  void InitMetricsReporter();
+
+  ///  It's periodic reporter entry for all runtime modules.
+  ///  \param metric_name, metric name
+  ///  \param value, metric value
+  ///  \param tags, metric tag map
+  void ReportMetrics(const std::string &metric_name, double value,
+                     const std::unordered_map<std::string, std::string> &tags = {});
+  void EnableTimer(std::function<void()> report_timer_handler);
```
Comment (MissiontoMars) [inline streaming/src/runtime_context.h:44]: Add more comments

File: streaming/src/runtime_context.cc [hunk 2]
```diff
@@ -26,7 +27,111 @@ void RuntimeContext::SetConfig(const uint8_t *data, uint32_t size) {
...
+    }
+    STREAMING_LOG(INFO) << "Timer server shutdown";
+    enable_timer_service_ = false;
+    STREAMING_LOG(INFO) << "Cancel metrics timer.";
+    metrics_timer_->cancel();
+  }
+  STREAMING_LOG(INFO) << "Wake up all reporting conditions.";
+  if (timer_thread_) {
+    STREAMING_LOG(INFO) << "Join and reset timer thread.";
+    timer_thread_->join();
```
Comment (MissiontoMars) [inline streaming/src/runtime_context.cc:130]: `if (joinable())`

File: streaming/src/runtime_context.cc [hunk 3]
```diff
@@ -26,7 +27,111 @@ void RuntimeContext::SetConfig(const uint8_t *data, uint32_t size) {
...
+    STREAMING_LOG(WARNING) << "metrics is disable";
+    return;
+  }
+  perf_metrics_reporter_.reset(new StreamingPerf());
+
+  std::unordered_map<std::string, std::string> default_tag_map = {
+      {"role", NodeType_Name(config_.GetNodeType())},
+      {"op_name", config_.GetOpName()},
+      // To avoid opentsdb query error if there are too many tags in metric.
+      /*{"pid", std::to_string(getpid())},*/
```
Comment (MissiontoMars) [inline streaming/src/runtime_context.cc:45]: Remove this comment

File: streaming/src/runtime_context.h [hunk 4]
```diff
@@ -30,9 +33,36 @@ class RuntimeContext {
...
+  void ShutdownTimer();
+
+ private:
+  void RunTimer();
+
+ protected:
+  std::unique_ptr<StreamingPerf> perf_metrics_reporter_;
+  std::function<void()> report_timer_handler_;
+
+  boost::asio::io_service async_io_;
```
Comment (MissiontoMars) [inline streaming/src/runtime_context.h:54]: a `boost::asio::work` should be added?

File: streaming/src/metrics/stats_reporter.h [hunk 5]
```diff
@@ -0,0 +1,74 @@
+#pragma once
```
Comment (MissiontoMars) [inline streaming/src/metrics/stats_reporter.h:1]: Can you add the license information on top of all files?

File: streaming/BUILD.bazel [hunk 6]
```diff
@@ -93,6 +100,24 @@ cc_library(
...
+    name = "streaming_metrics",
+    srcs = glob([
+        "src/metrics/*.cc",
+    ]),
+    hdrs = glob([
+        "src/metrics/*.h",
+    ]),
+    copts = COPTS,
+    strip_include_prefix = "src",
+    visibility = ["//visibility:public"],
```
Comment (MissiontoMars) [inline streaming/BUILD.bazel:113]: It doesn't need to be public.

File: streaming/src/metrics/streaming_perf_metric.h [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+#define METRIC_GROUP_JOIN(a, b, c) (a + "." + b + "." + c)
+
+class StreamingPerfBase {
+ public:
+  virtual ~StreamingPerfBase() = default;
+  virtual bool Start(const StreamingMetricsConfig &conf) = 0;
+
+  virtual void Shutdown() = 0;
+
+  virtual void UpdateCounter(const std::string &domain, const std::string &group_name,
```
[the remark this answers is not in the document]
Comment (ashione) [reply streaming/src/metrics/streaming_perf_metric.h:15]: > `UpdateXXX` -> `XXX` to simplify

Counter/Gauage looks like a constructor. So we just update it view data.

File: streaming/src/runtime_context.h (hunk 4 above)
Comment (ashione) [reply streaming/src/runtime_context.h:54]: boost::asio::work is not recommended after boost 1.67. In our case, we can make sure all of behaviors will be finished.

> a `boost::asio::work` should be added?

**Comment by jovany-wang** [comment]
Retrying...

**Review by jovany-wang — APPROVED** [review]

**Pull request merged**